searx/weather.py

```diff
@@ -0,0 +1,610 @@
+# SPDX-License-Identifier: AGPL-3.0-or-later
+"""Implementations used for weather conditions and forecast."""
+# pylint: disable=too-few-public-methods
+from __future__ import annotations
+
+__all__ = [
+    "symbol_url",
+    "Temperature",
+    "Pressure",
+    "WindSpeed",
+    "RelativeHumidity",
+    "Compass",
+    "WeatherConditionType",
+    "DateTime",
+    "GeoLocation",
+]
+
+import typing
+
+import base64
+import datetime
+import dataclasses
+
+from urllib.parse import quote_plus
+
+import babel
+import babel.numbers
+import babel.dates
+import babel.languages
+
+from searx import network
+from searx.cache import ExpireCache, ExpireCacheCfg
+from searx.extended_types import sxng_request
+from searx.wikidata_units import convert_to_si, convert_from_si
+
+WEATHER_DATA_CACHE: ExpireCache = None  # type: ignore
+"""A simple cache for weather data (geo-locations, icons, ..)"""
+
+YR_WEATHER_SYMBOL_URL = "https://raw.githubusercontent.com/nrkno/yr-weather-symbols/refs/heads/master/symbols/outline"
+
+
+def get_WEATHER_DATA_CACHE():
+
+    global WEATHER_DATA_CACHE  # pylint: disable=global-statement
+
+    if WEATHER_DATA_CACHE is None:
+        WEATHER_DATA_CACHE = ExpireCache.build_cache(
+            ExpireCacheCfg(
+                name="WEATHER_DATA_CACHE",
+                MAX_VALUE_LEN=1024 * 200,  # max. 200kB per icon (icons have most often 10-20kB)
+                MAXHOLD_TIME=60 * 60 * 24 * 7 * 4,  # 4 weeks
+            )
+        )
+    return WEATHER_DATA_CACHE
+
+
+def _get_sxng_locale_tag() -> str:
```

Comment:
Maybe this should be moved to https://github.com/searxng/searxng/blob/master/searx/locales.py? It's not necessarily coupled to the specific weather engine request and could be useful in other places too.

Reply:
We have several places in the code where we try to determine the "ideal" language selection for the client. All of these implementations still differ in detail today. This mess will have to be cleaned up one day, but until then, it's more reliable to implement this method here rather than, for example, this one (which does something similar).

https://github.com/searxng/searxng/blob/848c8d0544eb0138da4414a750f79b22883f91b2/searx/webapp.py#L161-L164

I'm working on a parallel branch that cleans up the properties and also takes this problem into account to find a general solution.